Kudu.Core/Deployment/DeploymentManager.cs

```diff
@@ -707,7 +707,12 @@ private DeployResult GetResult(string id, string activeDeploymentId, bool isDepl
                     {
                         await builder.Build(context);
                         builder.PostBuild(context);
-                        await RestartMainSiteIfNeeded(tracer, logger, deploymentInfo);
+
+                        // Don't request a second container restart if it is already being restarted as part of the PushDeploy workflow
+                        if (string.IsNullOrEmpty(deploymentInfo.DeploymentId))
+                        {
+                            await RestartMainSiteIfNeeded(tracer, logger, deploymentInfo);
+                        }
```

Comment:
Would be better to add this check for DeploymentId along with other checks inside RestartmainSiteIfNeeded. That way RestartmainSiteIfNeeded can be called unconditionally and as the name says - the site will be restarted only if needed.

Also, you may want to consider posting the restart operation to the FE from within RestartMainSiteIfNeeded. RestartmainSiteIfNeeded already has 2 ways of requesting a restart. One way is using a filesystem based notification. Second is using the /api/restart API. The new one being introduced is the third way and unless there is a good reason it would be good to do it from within RestartMainSiteIfNeeded. That way callers can call RestartMainSiteIfNeeded without having to worry about the details. Also - this keeps it consistent with KuduLite.

Reply:
Also, make sure that /api/publish?restart=false does not trigger a restart (neither the old way or new way of restarting)